lib/stages/BeamBufferSort.cpp

```diff
@@ -0,0 +1,295 @@
+#include <iostream>    // for cout
+#include <regex>       // for match_results<>::_Base_type
+#include <stdexcept>   // for runtime_error
+#include <string>      // for string
+#include <visUtil.hpp> // for frameID, modulo
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+using std::string;
+
+REGISTER_KOTEKAN_STAGE(BeamBufferSort);
+
+BeamBufferSort::BeamBufferSort(Config& config_, const std::string& unique_name,
+                               bufferContainer& buffer_container) :
+    Stage(config_, unique_name, buffer_container, std::bind(&BeamBufferSort::main_thread, this)) {
+    uint32_t nchan;
+    uint32_t buf_start_f = 0;
+    uint32_t buf_nchan;
+    // Apply config
+    samples_per_data_set = config.get<uint32_t>(unique_name, "samples_per_data_set");
+    has_freq_bin = config.get_default<bool>(unique_name, "has_freq_bin", false);
+    total_freq_chan = config.get_default<uint32_t>(unique_name, "total_freq_chan", 1024);
+    use_n_out_buffer = config.get_default<uint32_t>(unique_name, "use_n_out_buffer", total_freq_chan);    
+    wait_nframes = config.get_default<int>(unique_name, "wait_nframes", 20);
+    time_resolution = config.get_default<double>(unique_name, "time_resolution", 2.56e-6);
+    FreqIDBeamMeta_size = sizeof(FreqIDBeamMetadata);
+    // Total time for one coming frame in nanoseconds.
+    subframe_time_nsec = time_resolution * samples_per_data_set * 1000000000;
+    // Set up buffers
+    in_buf = get_buffer("in_buf");
+    register_consumer(in_buf, unique_name.c_str());
+    // Register the out buffer array
+    out_bufs = get_buffer_array("out_bufs");
+    for (struct Buffer* out_buf : out_bufs)
+        register_producer(out_buf, unique_name.c_str());
+    // Check if the request out buffer more than the registered buffers.
+    assert(use_n_out_buffer <= out_bufs.size());
+    
+    // Init sort queue
+    sub_frame_size = FreqIDBeamMeta_size + in_buf->frame_size;
+    //Init sort queue frequency axis
+    for (int i = 0; i < wait_nframes; ++i){
+        sort_queue.push_back((uint8_t*) malloc (sizeof(uint8_t) * total_freq_chan * sub_frame_size));
+	queue_status.push_back(std::vector<uint8_t> (total_freq_chan, 0));
+    }
+
+    // Assign frequency to the buffers.
+    nchan = total_freq_chan / use_n_out_buffer;
+    // Check if there is any extra channels
+    uint32_t rmder = total_freq_chan % use_n_out_buffer;
+    // Assign channels to the output buffer.
+    for (uint32_t i = 0; i< use_n_out_buffer; ++i){
+        if (i == 0){
+	    // First buffer gets more frequency chan
+	    buf_nchan = nchan + rmder;
+	}
+	else{
+	    buf_nchan = nchan;
+	}
+        out_buf_nchan.push_back(buf_nchan);
+	out_buf_frame0.push_back(0);
+	out_buf_start_chan.push_back(buf_start_f);
+	buf_start_f += buf_nchan;
+    }
+    
+}
+
+BeamBufferSort::~BeamBufferSort() {}
+
+void BeamBufferSort::make_freq_meta(FreqIDBeamMetadata* out_meta, uint32_t freq_id,\ 
```

Comment:
Minor point, I'd generally call this `fill_` function, since we aren't actually making/allocating new memory. 
```suggestion
void BeamBufferSort::fill_freq_meta(FreqIDBeamMetadata* out_meta, uint32_t freq_id,\ 
```